aptos-move/framework/aptos-stdlib/sources/cryptography/bulletproofs.move

```diff
@@ -0,0 +1,294 @@
+            proof.bytes,
+            num_bits,
+            dst
+        )
+    }
+
+    /// Verifies a zero-knowledge range proof that the value `v` encrypted by `ct` satisfies $v \in [0, 2^{num_bits})$. Only works
+    /// for `num_bits` \in {8, 16, 32, 64}.
+    public fun verify_range_proof_elgamal(ct: &elgamal::Ciphertext, proof: &RangeProof, pubkey: &elgamal::Pubkey, num_bits: u64, dst: vector<u8>): bool {
+        if(!features::bulletproofs_enabled()) {
+            abort(std::error::invalid_state(E_NATIVE_FUN_NOT_AVAILABLE))
+        };
+
+        verify_range_proof_custom_ck_internal(
+            ristretto255::point_to_bytes(&elgamal::get_value_component_compressed(ct)),
+	    &ristretto255::basepoint(), &elgamal::get_point_from_pubkey(pubkey),
+            proof.bytes,
+            num_bits,
+            dst
+        )
+    }
+
+    /// Verifies a zero-knowledge range proof that the value `v` committed in `com` (as v * val_base + r * rand_base,
+    /// for some randomness `r`) satisfies $v \in [0, 2^{num_bits})$. Only works for `num_bits` \in {8, 16, 32, 64}.
+    public fun verify_range_proof_custom_ck(
+        com: &pedersen::Commitment,
+        val_base: &RistrettoPoint, rand_base: &RistrettoPoint,
+        proof: &RangeProof, num_bits: u64, dst: vector<u8>): bool
+    {
+        if(!features::bulletproofs_enabled()) {
+            abort (std::error::invalid_state(E_NATIVE_FUN_NOT_AVAILABLE))
+        };
+
+        verify_range_proof_custom_ck_internal(
+            ristretto255::point_to_bytes(&pedersen::commitment_as_compressed_point(com)),
+            val_base, rand_base,
+            proof.bytes, num_bits, dst
+        )
+    }
+
+    /// Verifies a zero-knowledge range proof that the value `v` encrypted by `ct` (as v * val_base + r * rand_base,
+    /// for some randomness `r`) satisfies $v \in [0, 2^{num_bits})$. Only works for `num_bits` \in {8, 16, 32, 64}.
+    public fun verify_range_proof_custom_valbase_elgamal(
+        ct: &elgamal::Ciphertext,
+        val_base: &RistrettoPoint, rand_base: &RistrettoPoint,
+        proof: &RangeProof, num_bits: u64, dst: vector<u8>): bool
+    {
+        if(!features::bulletproofs_enabled()) {
+            abort (std::error::invalid_state(E_NATIVE_FUN_NOT_AVAILABLE))
+        };
+
+        verify_range_proof_custom_ck_internal(
+            ristretto255::point_to_bytes(&elgamal::get_value_component_compressed(ct)),
+            val_base, rand_base,
+            proof.bytes, num_bits, dst
+        )
+    }
+
+    #[test_only]
+    /// Computes a range proof for the Pedersen commitment to 'val' with 'randomness', under the default Bulletproofs
+    /// commitment key; see `pedersen::new_commitment_for_bulletproof`. Returns the said commitment too.
+    ///  Only works for `num_bits` \in {8, 16, 32, 64}.
+    public fun prove_range(val: &Scalar, r: &Scalar, num_bits: u64, dst: vector<u8>): (RangeProof, pedersen::Commitment) {
```

Comment:
This will now need to be `prove_range_pedersen`.

And, AFAICT, we will need another `prove_range`, which calls a `prove_range_internal` native, which takes a custom `val_base` and `rand_base` as input.

Then, we can add a `prove_range_elgamal` which just calls `prove_range` with the right bases.

And, we can also refactor `prove_range_pedersen` to do the same.

This way, there is only one `prove_range` native.